### ISSUE TITLE
Issue with running custom density and abundance profile

### ISSUE THREAD

**Issue by manolis07gr** [tardis-sn/tardis#722]
Dear all,

I am trying to run a properly formatted TARDIS spectrum model for a given custom density and abundance profile taken from an 1D radiation hydrodynamics simulation. Attached there is a zipped file of the csm_sn.yml file and the custom abundance and density profile models I am using. I keep getting the following error:

tardis.simulation.base - INFO - Luminosity emitted = 0.00000e+00 erg / s Luminosity absorbed = 3.84573e+42 erg / s Luminosity requested = 3.84600e+42 erg / s
Traceback (most recent call last):
  File "/Users/macbro/anaconda3/envs/tardis/bin/tardis", line 75, in <module>
    run_radial1d(radial1d_mdl)
  File "/Users/macbro/anaconda3/envs/tardis/lib/python2.7/site-packages/tardis/simulation/base.py", line 433, in run_radial1d
    hdf_mode, hdf_last_only)
  File "/Users/macbro/anaconda3/envs/tardis/lib/python2.7/site-packages/tardis/simulation/base.py", line 268, in legacy_run_simulation
    self.tardis_config.supernova.luminosity_requested)
  File "/Users/macbro/anaconda3/envs/tardis/lib/python2.7/site-packages/tardis/simulation/base.py", line 91, in estimate_t_inner
    return input_t_inner * luminosity_ratios ** t_inner_update_exponent
ZeroDivisionError: 0.0 cannot be raised to a negative power

I am not sure what I am missing or if I am still setting some parameters incorrectly. I would appreciate your help with producing a model spectrum with this!

Best,

Dr. Emmanouil (Manos) Chatzopoulos
Assistant Professor
Nicholson 275
Department of Physics & Astronomy
Louisiana State University

[csm.tar.gz](https://github.com/tardis-sn/tardis/files/906808/csm.tar.gz)

**Comment by yeganer** [comment]
If I read the log correctly, then no radiation escapes the photosphere,
My guess is that the density is specified in the wrong units OR there is a bug in the code. However I think reading the configuration is part of our integration tests.

This is a sample file, of how the structure should look like:
[https://github.com/tardis-sn/tardis/blob/master/tardis/tests/integration_tests/w7/densities.dat](https://github.com/tardis-sn/tardis/blob/master/tardis/tests/integration_tests/w7/densities.dat)

I hope, this helps

**Comment by unoebauer** [comment]
After looking at the config and density files you have attached, I am quite sure that @yeganer pointed into the right direction. For 76 days, the density profile you specified seems to be much too high. A rough integration of the profile yields an ejecta mass of ~1e16 Msun!

Tardis will read in the density file, interpret it as providing the profile at the time specified in the first line of the file and then homologously expand (compress) it to the time since explosion specified in the yml config file. Since in your example these two times are identical (intentional?), no expansion is applied and the very dense profile specified in the file is used together with a very large ejecta (hence the incredibly high optical depth and the complete blocking of all photospheric radiation).

**Comment by unoebauer** [comment]
I am also wondering if it was your intention to place the photosphere at vphot = 32 km/s? Provided that you fix the density issue I suspect that the spectrum will still look quite odd. Being located so deep inside the ejecta, the photospheric surface is quite small compared to the overall size of the ejecta. The only way to accommodate the emergent luminosity request is by imposing a very high photospheric temperature, which in turn will lead to very high plasma temperatures in the ejecta.

If you want to move the photosphere further out, you can simply add back the v_inner_boundary parameter in the config file. Tardis will then use the parts of the density and abundance structure specified in the files which lie above this boundary.

I hope these remarks help. Let us know if you have additional questions. You can also contact us directly on our gitter channel: https://gitter.im/tardis-sn/tardis

**Comment by manolis07gr** [comment]
[csm.tar.gz](https://github.com/tardis-sn/tardis/files/914379/csm.tar.gz)

Dear all,

First of all thank you so much for your helpful reply and insights! I had a bug in my python code converting the (Eulerial) radiation hydrodynamics profiles I had into the file formats required by TARDIS after I "homologize" the SN ejecta (plus circumstellar matter around the SN ejecta where the collision takes place and where spectrum formation occurs). I attach the updated files (with correct densities now!) together with the output spectrum (that still looks odd to me). As a cautionary note this was run using the same reference time in the density file and in the input (yaml file) which, as you pointed out it was wrong. That was confusing me initially as I thought that in the reference time in the density file I should be using the same "phase" that I am trying to post-process from the hydro simulation. 

In other words I was trying to do a "snapshot" post-processing type calculation. I assume the right thing to do would have been to use an earlier profile with the right reference time and then request the corresponding later time I wanted to compute using time_explosion in the yaml file? Or, could I alternatively set the output profile reference time to 1 day and then request a time_explosion only slightly higher (like, 1.1) day so that TARDIS effectively calculates the spectrum almost at the same phase?

Any input on that would be greatly appreciated. Thanks!

Manos

**Comment by unoebauer** [comment]
Looking at the files you've attached, in particular at the spectrum, it looks as if my second suspicion came true. Since you did not specify a ``v_inner_boundary`` in the yml file, TARDIS set the photosphere to the innermost cell of the discretization specified in the density file (i.e. to 39km/s). Since you requested a very high output luminosity and since the photosphere is so small, TARDIS converged to a very high photospheric temperature to achieve this emergent luminosity. Hence the very blue spectrum and the overall peculiar appearance. Since TARDIS does not include continuum process in its current implementation, you should not set the photosphere too deep inside envelope. Instead, it should be at the bottom of the line forming region. As you mentioned in your post, you expect line formation to happen in the CSM shell. In this case you should move the photosphere out to these regions (via the ``v_inner_boundary`` parameter in the yml file).

**Comment by unoebauer** [comment]
Concerning the density profile it is quite straightforward: in the density file itself, you specify the time since explosion corresponding to the profile defined in the file. In the yml config file you give the time since explosion for which you want TARDIS to calculate a spectrum. Any difference between these two times is accounted for internally by TARDIS via a homologous expansion/compression which is applied to the density profile.

(**One caveat**: TARDIS currently does not track radioactive decays. Thus, you have to decay Ni56, Co56, etc. by hand yourself if you want to have the appropriate elemental composition. See [TEP007](https://github.com/tardis-sn/tep/blob/master/TEP007_isotopes_decay.rst) )

**Comment by unoebauer** [comment]
One final remark with respect to the density profile you set up: it has a very peculiar density spike in the last, outermost cell. Here the density is ~1e2 times higher than in the adjacent cells. Thus, the cell will contain quite a bit of mass and will most likely dominate spectra formation in TARDIS. This seems a bit odd to me...

**Comment by manolis07gr** [comment]
Hi Unoebauer,

Thanks for your helpful explanation on the reference time-scales. As far as your question goes on the outer density spike I wanted to say that this is set there in purpose (I expect the photosphere to be out there unless I should also add some layers on top of that). This profile is taken by a simulation in which the SN ejecta rammed and interacted with a circumstellar (CSM) shell. The spike is the thin dense shell forming bounded by the forward shock (moving into the CSM - in mass coordinates) and the reverse shock (moving into the SN ejecta - in mass coordinates) while all the "double-shock" structure moves outwards in radius. So this dense shell is where most of the mass indeed is and where I expect the line-forming region and the photosphere to be. Do you think I should modify it or perhaps resolve it better?

Also, when I enable the v_inner_boundary flag in my yaml file I always get the following error:

Traceback (most recent call last):
  File "/Users/manolis/anaconda2/envs/tardis/bin/tardis", line 65, in <module>
    tardis_config = config_reader.Configuration.from_yaml(args.config_fname)
  File "/Users/manolis/anaconda2/envs/tardis/lib/python2.7/site-packages/tardis/io/config_reader.py", line 747, in from_yaml
    yaml_dict, *args, **kwargs)
  File "/Users/manolis/anaconda2/envs/tardis/lib/python2.7/site-packages/tardis/io/config_reader.py", line 783, in from_config_dict
    validated_config_dict = config_validator.validate_dict(config_dict)
  File "/Users/manolis/anaconda2/envs/tardis/lib/python2.7/site-packages/tardis/io/config_validator.py", line 72, in validate_dict
    ).validate(validated_dict)
  File "/Users/manolis/anaconda2/envs/tardis/lib/python2.7/site-packages/jsonschema/validators.py", line 123, in validate
    raise error
jsonschema.exceptions.ValidationError: OrderedDict([('type', 'file'), ('filename', 'pm2_dens60b.dat'), ('filetype', 'simple_ascii'), ('v_inner_boundary', 2200)]) is not valid under any of the given schemas

Failed validating 'oneOf' in schema['properties']['model']['properties']['structure']:
    OrderedDict([('oneOf', [OrderedDict([('$ref', '#/definitions/structure/file')]), OrderedDict([('$ref', '#/definitions/structure/specific')])])])

On instance['model']['structure']:
    OrderedDict([('type', 'file'), ('filename', 'pm2_dens60b.dat'), ('filetype', 'simple_ascii'), ('v_inner_boundary', 2200)])

What am I doing wrong? The yaml file looks like it's formatted the right way. Here I put v_inner_boundary = 2200 (km/s) corresponding to the outer shell (in velocity coordinates) in the pm2_dens60.dat file.

Thanks again for your continuing help!

Manos

**Comment by manolis07gr** [comment]
Oh I see, the units where missing from v_inner_boundary (km/s). Now I set: 

        v_inner_boundary: 100 km/s  # WAS 11000 km/s
        v_outer_boundary: 2205 km/s # WAS 20000 km/s

Since setting v_inner_boundary to be as high as the v in the outer shell in the density file wouldn't run, you think these are appropriate choices for these two models?

And just to clarify, I am continuing by setting the reference time in the density file equal to what it was (taken from the original radiation hydro simulation) and then I can set explosion_time in yaml to re-scale the profile earlier/later in time.
As far as the luminosity_requested option goes I am choosing these values based on the luminosity that I find at the photosphere of this profile (from the rad hydro simulation) at this same phase (ref time I input on top of the density file). 
Let me know if you still think my approach is not crazy. Thanks again!

Best,

Manos

**Comment by unoebauer** [comment]
Hi @manolis07gr - to clarify, the ``luminosity_requested`` is actually interpreted by TARDIS as the luminosity emerging from the top of the ejecta/atmosphere. In general, this will be lower than the luminosity which is injected at the photosphere, since part of the radiation is backscattered onto the photosphere and thus lost. In the plasma iterations TARDIS will try to adjust the temperature at the photosphere in a way that enough radiation escapes from the top of the ejecta to meet the requirement. 

Given the handling of ``luminosity_requested``, I don't think that setting it to the value you get from the RH simulation is the best thing to do. In your case, it may be worth considering to instead incorporate the following parameters into your setup:
```
montecarlo:
  convergence_strategy:
    type: damped
    t_inner:
      damping_constant: 0.0
plasma:
  initial_t_inner: 18640 K
```
This keeps the photosphere at a constant temperature, ``initial_t_inner``. You should replace the ``initial_t_inner`` value with the value you find in your RH simulation at the photosphere. You should then also place the photosphere there (in terms of ``v_inner_boundary``). The region above the photosphere should be reasonably well resolved since this is the only part of the setup in TARDIS which can influence/alter the injected radiation field at the photosphere.

I should also add that your interaction problem is not the typical use case for which TARDIS was developed. So you are entering new territory in a sense...

Hope this helps

**Comment by manolis07gr** [comment]
Hi Uno,

Then I am still confused as to what I should set luminosity_requested to? Do I just comment it out? I try that and get (of course) a very dim (flat, no features) spectrum. I also added your suggested parameters in addition.

I know this is new territory but my hope is that TARDIS may shed some light into spectra coming from interaction. One feature I am aiming to get is a non-LTE hydrogen recombination line (intermediate width) that represents what happens to hydrogen-rich material shocked following SN ejecta - CSM interaction. Line is produced when H quickly recombines. Do you think TARDIS is able to produce H (Balmer) emission lines?

Thanks again,

Manos

**Comment by ssim** [comment]
Hi Manos

Just to respond on a couple of points there. I think it's correct that you should just remove (comment out) the "luminosity_requested" variable if you specify the inner boundary temperature instead (Uli or Stefan will know this better than me, however).

Unfortunately, the current version of TARDIS will not do a good job of simulating the H Balmer emission lines: specifically, emission due to recombination is not currently included in the calculation, and this will be a very important part of the emission. There are plans to to include this physics in TARDIS in due course, but it's not in the current version. This means that, at the moment, you'll only see H lines appearing if they are formed by scattering in the atmosphere, but it sounds like that's not the regime of primary interest?

Stuart

**Comment by manolis07gr** [comment]
Hi Stuart,

Good to hear from you and thank you on these points. Yes, recombination will be a dominant process happening in the context of strong SN Ejecta - CSM interaction and ideally it would be great to be able to model this with TARDIS in the near future though I am not quite sure of the code development changes that need to go along with that. Currently I am trying to use a mix between TARDIS and CLOUDY to get a pseudo-composite spectrum from both codes spanning both regimes (CLOUDY can do recombination lines but it suffers in higher densities characteristic of early SN ejecta and CSM interaction where TARDIS is stronger). Basically the idea would be to get underlying spectrum due to the SN by TARDIS and then convolve it with the CSM-dominated spectrum (H-lines) produced by CLOUDY for the time being through this approach is not fully self-consistent it might allow the investigation of the relevant parameter space. 

I did comment out the luminosity_requested variable by using the inner boundary temperature variable as suggested by Uli but this produced a very dim, (straight line, no features) wrong spectrum and I think it assumed by default a very low luminosity.

Thanks again,

Manos

**Comment by unoebauer** [comment]
Hi Manos,

as long as you've set ``damping_constant: 0.0`` and  ``initial_t_inner``, TARDIS has no freedom with respect to choosing the input luminosity at the photosphere. It will be:

```python
L_phot = 4 * pi * (v_inner_boundary * time_since_explosion)**2 * \
     pi * sigma_stefan_boltzmann * initial_t_inner**4
```

Due to back-scattering onto the photosphere, not all of this input luminosity will be able to escape from the top of the atmosphere. However, if there is a significant difference between input and emergent luminosity ( >~ factor 10), my conclusion would be that the photosphere is still too far inside (for the assumed photospheric temperature).

Status lines printed by TARDIS such as 
```
tardis.simulation.base - INFO - Luminosity emitted = 1.24608e+43 erg / s 
Luminosity absorbed = 4.00875e+42 erg / s 
Luminosity requested = 1.05928e+43 erg / s
```
will tell you about this backscattering. If you've set ``damping_constant: 0.0`` and  ``initial_t_inner``, you should ignore ``Luminosity requested`` since it has no effect at all. The interesting numbers are ``Luminosity emitted``, which states how much energy is injected at the photosphere, and ``Luminosity absorbed``, which tells you how much radiation energy is backscattered onto the photosphere (and thus lost).

Uli

**Comment by manolis07gr** [comment]
Dear all,

Sorry to bother you again with this issue but I had yet another question with regards to the role of v_inner_boundary and v_outer_boundary inputs that I want to understand better. When I convert my rad hydro model to the TARDIS friendly format (density.txt) as discussed above, I am only using data up to optical depth of 0.001 but I also track the location of the photosphere at optical depth 0.3333. In other words, I have the entire SN ejecta imported up to tau = 0.001 and the region between tau = 0.333 (photosphere) and tau = 0.001 is just "fluff", or some low-density wind material in which the SN ejecta expands. Note: this new model only has the SN ejecta and not a CSM shell in the front, so it should be more straightforward.

You have been referring on using v_inner_boundary as the location of the photosphere in terms of velocity coordinates. But then since the maximum velocity is close to the location of the photosphere (in the homologous profile) that leaves only a tiny amount of (wind) material outside of v_inner_boundary (if coincided with the location of the photosphere). Why is the rest of the data (density profile within the star, interior to the photosphere) not used by TARDIS? Also, then, what is the role of v_outer_boundary?

Apologies if my question is naive but I want to make sure I understand this correctly as I run my models. If it helps, I am attaching the new test_dens50.dat file I am trying to run, based on your previous suggestions, that has its photosphere at 6640.68485 km/s leaving only 1 data point in the profile (in shell 49 in test_dens50.dat attached) outside of it. I also attach the corresponding abundance file as well as the csm_sn.yml TARDIS input file I am using based on your prior suggestions. 

Thank you,

Manos

**Comment by manolis07gr** [comment]
Ooops, forgot the attachment! Here it is
[newSNdata.tar.gz](https://github.com/tardis-sn/tardis/files/944601/newSNdata.tar.gz)

**Comment by unoebauer** [comment]
Hi Manos,

from a pure technical standpoint, ``v_inner_boundary`` and ``v_outer_boundary`` simply define the extent of the computational domain TARDIS works on. At v_inner_boundary radiation will be injected into the domain with a thermal frequency distribution and all radiation escaping through the outer boundary will be recorded and used to determine the emergent spectrum. 

From a physical standpoint all this goes back to the elementary supernova model (Jeffery & Branch 1990), which TARDIS adopts to a certain extent (as does SYN++ and the MC code by Mazzali & Lucy 1993). In this model, the spectrum formation process in supernovae is simplified by assuming that the ejecta can be separated into two regimes; namely an inner region in which energy generation happens, continuum processes play a role but line interactions are negligible and an atmosphere in which line formation takes place and continuum processes are unimportant. Both regions are separated by the photosphere at which radiation streams into the line formation region (atmosphere) with a thermal spectral energy distribution. TARDIS only treats the atmosphere region explicitly and does not include any continuum interaction processes therein. Thus, the photosphere is the only source of radiation energy. In the atmosphere, photons can only scatter around. As a consequence, this approach is only applicable as long as not too much radiation is lost (due to backscattering onto the photosphere) and as long as energy generation is not expected to happen in the atmosphere region (i.e. there should be no significant amounts of radioactive nickel above the photosphere). This means that the photosphere should not be placed too deep inside the ejecta (at high optical depths) since then the negligence of continuum process is no longer justified and too much radiation is backscattered onto the photosphere. Analogously, the photosphere should not be placed too far out (at very low optical depths) since then important parts of the line formation region are ignored.

Originally TARDIS was designed as a tool for abundance tomography purposes or simple fitting of observational data. In such applications, the location of the photosphere would be adjusted until the photospheric velocity and temperature (together with the assumed composition and density profile in the atmosphere) lead to an emergent spectrum which fits the observation.

I hope this helps. Let me/us know if some/all points should be elaborated

Uli

P.S.: I can't open your tar.gz. ``not in gzip format``

**Comment by manolis07gr** [comment]
[SNejecta.tar.gz](https://github.com/tardis-sn/tardis/files/947568/SNejecta.tar.gz)

Hi Uli,

Thank you so much for the detailed description of the code and the parameters. I was just naively confused as to why one needs to extract and upload an entire SN ejecta profile when TARDIS given the v_inner_boundary variable virtually ignores the stuff deep inside. I think you cleared up my confusion.

I have re-attached the SNejecta.tar.gz file here (re-compressed it for you) just in case you can take a look and see if what I get looks reasonable (again this is just for regular Type II ejecta from a 15Msun star explosion). 

I can unpack with: tar -xvf SNejecta.tar.gz

Thanks again,

Manos

**Comment by unoebauer** [comment]
I am closing this since there is nothing fundamentally wrong with the custom density and abundance profiles and since the misconceptions have been cleared up during the conversion.
